### ISSUE TITLE
GRPO: Why does loss start at 0 for first K steps and then increase over time?

### ISSUE THREAD

**Issue by arnavgarg1** [huggingface/trl#2703]
### Reproduction

Hi all! 

I've been trying to train a variety of models using GRPO, but I noticed that the train/loss metric remains 0 or close to 0 throughout training even after a large number of steps (>200). My mean rewards also don't change significantly during this same period. This seems unexpected and might indicate that the optimization step isn't working as intended.

However, on the official docs page, I see that that the image of the learning curves shared suggests that loss can remain close to 0 for long periods of time and actually increases instead of decreasing: https://huggingface.co/docs/trl/main/en/grpo_trainer#trl.GRPOConfig. Similarly, in @philschmid's blogpost from earlier today (https://www.philschmid.de/mini-deepseek-r1), I noticed a similar trend where loss stayed 0 for nearly 200 steps before increasing. 

This makes it seem like it is the expected behavior, but I'm having a hard time understanding it.

I had a few questions that I am hoping someone is able to help me understand:
1. Is there an issue with the loss computation in GRPO?
2. Should loss values be expected to remain near zero in this setting? What does an increasing loss suggest?
3. Could this be related to specific hyperparameters or gradient updates not being applied correctly?

Would appreciate any insights or guidance on debugging this! Thanks!

### System Info

- OS: Linux
- Transformers 4.48.1
- TRL: Main Branch

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

**Comment by qgallouedec** [comment]
Intereting question!

The answer is in the math. If you calculate the value of the loss (ignore the gradient), you'll see that it's equal to $\beta \mathrm{KL}$. That's why it starts at 0 and that's why it's incresing.

**Comment by arnavgarg1** [comment]
Thanks for the prompt response @qgallouedec! 

Does this mean that the loss itself is not a reliable indicator of training progression and we should primarily rely on KL and reward trends instead?

**Comment by qgallouedec** [comment]
You should rely mostly on the reward. And keep an eye on the generations (risk of reward hacking)

**Comment by NickyDark1** [comment]
I trained this model:
https://huggingface.co/NickyNicky/Llama-1B-base-GRPO-miniThinky_v1

I have these metrics train:

![Image](https://github.com/user-attachments/assets/e0e99b3d-4bec-4cbe-8eac-b03fef5689a1)

I see that it is not necessary to wait long for it to come out with a value of 0, I also observe that the sudden changes where the range from 300 to 500 tokens for generation and if it increases more say to 1000 tokens to generate only that change could reach the rewards to zero again and without leaving there

**Comment by NickyDark1** [comment]
other model train:

![Image](https://github.com/user-attachments/assets/fa4624ec-c083-40fb-99c7-e3467b14ace6)

![Image](https://github.com/user-attachments/assets/074c1950-46b9-48dc-b1dc-1ba174ad1f89)

I think token generation changes affect

**Comment by Jerrrrykun** [comment]
> Interesting question!
> 
> The answer is in the math. If you calculate the value of the loss (ignore the gradient), you'll see that it's equal to β KL . That's why it starts at 0 and that's why it's increasing.

Hi. I am a bit confused. From the implementation about `per_token_loss`, it should be `advantages` as `torch.exp(0)=1`. So how could it be `\beta KL` for a mount of steps?🤔

https://github.com/huggingface/trl/blob/af4ad47035529164799be10f3fe558ee642a9880/trl/trainer/grpo_trainer.py#L567
=

**Comment by XiaofengZHOU** [comment]
> > Interesting question!
> > The answer is in the math. If you calculate the value of the loss (ignore the gradient), you'll see that it's equal to β KL . That's why it starts at 0 and that's why it's increasing.
> 
> Hi. I am a bit confused. From the implementation about `per_token_loss`, it should be `advantages` as `torch.exp(0)=1`. So how could it be `\beta KL` for a mount of steps?🤔
> 
> #
> [trl/trl/trainer/grpo_trainer.py](https://github.com/huggingface/trl/blob/af4ad47035529164799be10f3fe558ee642a9880/trl/trainer/grpo_trainer.py#L567)
> 
> Line 567 in [af4ad47](/huggingface/trl/commit/af4ad47035529164799be10f3fe558ee642a9880)
> 
>  per_token_loss = torch.exp(per_token_logps - per_token_logps.detach()) * advantages.unsqueeze(1)

I think it's not the same as the original GRPO algorithm(missing ration and clamp)

**Comment by qgallouedec** [comment]
> I think it's not the same as the original GRPO algorithm(missing ration and clamp)

It is the same, since we do 1 optimization step

**Comment by XiaofengZHOU** [comment]
> > I think it's not the same as the original GRPO algorithm(missing ration and clamp)
> 
> It is the same, since we do 1 optimization step

according to the equation，the loss== βKL, which means the bigger the kl, the better the performance?
so how does the reward work?

for example:
rewards = torch.tensor([ 0, 1, 0], dtype=torch.float32)
per_token_logps1 = [[-0.4, -0.3], [-0.6, -0.5], [-1, -1]]
per_token_logps2 = [[-0.6, -0.5], [-0.4, -0.3], [-1, -1]]
the loss calculated are bot tensor(0.0062).

can you point out where is my problem?

**Comment by zhangsheng377** [comment]
@qgallouedec Now I understand why the loss of grpo start from 0. But there is another question, why can training still proceed after the loss starts from 0? Is it because the GPU accumulates computational errors, resulting in KL not being 0, thus allowing the formal training to begin?

**Comment by artshcherbina** [comment]
Even if loss is 0, the loss gradient is not. Thus the training proceeds

**Comment by Cppowboy** [comment]
If I set \beta to 0 and use sequence mean loss, the loss will always be 0?

**Comment by Cppowboy** [comment]
> > Interesting question!
> > The answer is in the math. If you calculate the value of the loss (ignore the gradient), you'll see that it's equal to β KL . That's why it starts at 0 and that's why it's increasing.
> 
> num_generations output o1~o_g，have different length, so, for every A_it， have been accumulated different times, it's not equal to β KL

So GRPO is a length biased algorithm?

**Comment by Rakshith12-pixel** [comment]
@qgallouedec attached below is my loss curve for GRPO on qwen2-vl7b instruct model, on which I had already performed SFT on my task

<img width="1157" height="674" alt="Image" src="https://github.com/user-attachments/assets/680ad210-43fc-41f3-afc5-1c8f9158b2e8" />

I notice that the loss doesn't start from zero. Are these curves healthy?? the KL curve increases like how the others have posted

**Comment by qgallouedec** [comment]
Which loss type do you use? What is the value of beta?
If you use the default config, then it's perfectly expected

**Comment by Rakshith12-pixel** [comment]
Thanks for your response

I just kept the default config.
But then why aren't my rewards increasing?? I did eval on the checkpoint after 0.75 epochs and there is no improvement on OCR performance.

Pardon me for my naivety , though I have decent knowledge about the algorithm, this is the first time im training GRPO

**Comment by qgallouedec** [comment]
It's hard to tell with no dataset, config, model, reward implementation etc? I'm going to close this issue because the original question was solved.

There is a detailed explanation here as well: https://github.com/huggingface/open-r1/issues/239#issuecomment-2646297851
